### DISCUSSION TITLE
How to get -level to behave as expected on alpha channel?

### DISCUSSION THREAD

**Discussion by flintstryker** [ImageMagick/ImageMagick#2000] [General]
I'm trying to use ImageMagick to process a lot of icons that have unsightly semitransparent pixels. I need SOME semitransparency for anti-aliasing, so I need to adjust the alpha channel levels so that the faintest pixels become fully transparent, and the almost-opaque pixels become fully opaque, but still preserve the values in between. The command I'm running looks like this:

magick mogrify -channel Alpha -level 40,215 *.png

The file I'm working with looks like this:
![1588982541](https://user-images.githubusercontent.com/65050612/81458127-7d198800-915e-11ea-988a-83614bb2b023.png)

What I expect to happen is that any alpha values under 40 become 0, any values over 215 become 255, and the rest of the values are spread between 0 and 255.
This is what I expect the output to look like:
![image](https://user-images.githubusercontent.com/65050612/81458142-8c003a80-915e-11ea-8996-9934fa02da1e.png)

Instead, the results that I'm getting look like the command is just forcing all alpha values >0 to become 255.
This is what I'm getting:
![1588982661](https://user-images.githubusercontent.com/65050612/81458191-c2d65080-915e-11ea-8fd9-c8a6d49b16a6.png)

I've tried numerous variations of the command; tired using percentages, decimals, added a gamma value, used only a single value (black_point value only), reversed the values, and banged my head on the wall in other ways. I even tried -level 0,255 which in theory shouldn't do anything, but got the same result.

**Comment by fmw42** [comment]
try using -black-threshold and -white-threshold on the alpha channel. See https://imagemagick.org/script/command-line-options.php#black-threshold
https://imagemagick.org/script/command-line-options.php#white-threshold

**Reply by flintstryker** [reply]
Thanks. I gave black-threshold a try. All the values I've tried (0%, 10%, 90%, 100%) give the same result:
![1588983682](https://user-images.githubusercontent.com/65050612/81458677-3f6a2e80-9161-11ea-8d18-1e476638bcc9.png)

I'm reinstalling ImageMagick now in case this is just a weird bug. Any other ideas why this is happening?

**Reply by fmw42** [reply]
Post your original image. What you posted has a checkerboard background. We need to see the original file.

**Reply by snibgo** [reply]
> magick mogrify -channel Alpha -level 40,215 *.png

What version of IM do you use, on what platform? Critically, what is the Q-number: Q8, Q16, Q32?

Your command assumes Q8, so normal values are in the range 0 to 255. But I suspect you have Q16 so values are in the range 0 to 65535. The highest normal value is QuantumRange.

I suggest you write commands that are independent of the Q-number, so they will always work. For example:

> magick mogrify -channel Alpha -level %[fx:40\*QuantumRange/255],%[fx:215\*QuantumRange/255] *.png

**Reply by fmw42** [reply]
Use percent in -black-threshold X% and in the same command -white-threshold Y% where X is small and Y is large.

`magick image -channel a -black-threshold X% -white-threshold Y% +channel result`

**Reply by flintstryker** [reply]
Thanks for the help guys. My problem was twofold: 

1. I'm using the Windows binary version 7.0.10-10-Q16-x64. @snibgo you were correct. I had no idea that QuantumRange was based on the version!

2. Not wanting to deal with variables in the command, I decided to revisit using percents as suggested by @fmw42. Then it dawned on me: In my frustration had I forgotten that Windows batch files must use %% instead of %. Replacing % with %% gives the correct output.
